### PR TITLE
Fix align reconstructions and re-compute UID option

### PR DESCRIPTION
**Pull request by simogasp**
* [X] fix bug connected to poparteu/openMVG#201
* [X] add an option to `ConvertSfM_DataFormat` to re-compute the UID, useful in case of legacy reconstructions that were not using UIDs.

Other spur of the moment fix/features
* [X] also fix cmake bug, using `BOOST_LIBRARIES` instead of `Boost_LIBRARIES` connected to poparteu/openMVG#203
* [X] some refactoring (mostly renaming) of `LocalizationResults`
  * also fix unity test as `indMatch3D2D` have must the same size as matched points
* [X] bug fix, control point were not transformed (connected to poparteu/openMVG#207)
* [X] bug fix, always save LocalizationResults in cameraLocalizer (connected to poparteu/openMVG#209)

### CODE REVIEW

File: src/openMVG/localization/LocalizationResult.hpp [hunk 1]
```diff
@@ -80,6 +79,8 @@ class LocalizationResult {
   {
     return _matchData.projection_matrix;
   }
+  
+  sfm::Image_Localizer_Match_Data getMatchData() const { return _matchData; }
```
Comment (fabiencastan) [inline src/openMVG/localization/LocalizationResult.hpp]: `const sfm::Image_Localizer_Match_Data&`

File: src/openMVG/localization/LocalizationResult.hpp [hunk 2]
```diff
@@ -163,23 +165,28 @@ class LocalizationResult {
   }
   
 private:
+  
+  // Hold all the imaged points, their associated 3D points and the inlier indices
```
Comment (fabiencastan) [inline src/openMVG/localization/LocalizationResult.hpp]: Use `///` to get it in the doxygen documentation.

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp [hunk 3]
```diff
@@ -5,24 +5,131 @@
...
+#include <boost/filesystem.hpp>
+#endif
+
 #include <string>
 #include <vector>
 
 using namespace openMVG;
 using namespace openMVG::sfm;
 
+void updateStructureWithNewUID(Landmarks &landmarks, const std::map<std::size_t, std::size_t> &oldIdToNew)
```
Comment (fabiencastan) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]:
```
/**
 * @brief Change all viewID referenced in landmarks.
 */
void updateViewIDs(Landmarks &landmarks, const std::map<std::size_t, std::size_t> &oldIdToNew)
```

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp [hunk 4]
```diff
@@ -5,24 +5,131 @@
...
+
+      newObservations.emplace(oldIdToNew.at(idview), obs);
+    }
+    
+    assert(currentLandmark.obs.size() == newObservations.size());
+    currentLandmark.obs.swap(newObservations);
+  }  
+}
+
+void regenerateUID(sfm::SfM_Data &sfmdata, std::map<std::size_t, std::size_t> &oldIdToNew, bool sanityCheck = false)
```
Comment (fabiencastan) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]:
```
/**
 * @brief Recompute the UID from the metadata of the original input images and modify the ID if it's not the same.
 */
```

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp [hunk 5]
```diff
@@ -84,9 +204,53 @@ int main(int argc, char **argv)
...
+    regenerateUID(sfm_data, oldIdToNew);
+    
+#ifdef HAVE_BOOST
+    if(!matchDir.empty())
+    {
+      std::cout << "Generating alias for .feat and .desc with the UIDs" << std::endl;
+      for(const auto& iter : oldIdToNew)
+      {
+        const auto oldID = iter.first;
+        const auto newID = iter.second;
```
Comment (fabiencastan) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]: Need to continue if `oldID == newID`.

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp [hunk 6]
```diff
@@ -31,7 +138,11 @@ int main(int argc, char **argv)
   cmd.add(make_switch('S', "STRUCTURE"));
   cmd.add(make_switch('O', "OBSERVATIONS"));
   cmd.add(make_switch('C', "CONTROL_POINTS"));
+  cmd.add(make_switch('u', "uid"));
```
Comment (fabiencastan) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]: `cmd.add(make_switch('u', "regenerateUID"));`

File: src/openMVG/localization/LocalizationResult.hpp (hunk 1 above)
Comment (simogasp) [inline src/openMVG/localization/LocalizationResult.hpp]: done

File: src/openMVG/localization/LocalizationResult.hpp (hunk 2 above)
Comment (simogasp) [inline src/openMVG/localization/LocalizationResult.hpp]: done

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp (hunk 3 above)
Comment (simogasp) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]: done

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp (hunk 4 above)
Comment (simogasp) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]: done

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp (hunk 6 above)
Comment (simogasp) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]: done

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp (hunk 5 above)
Comment (simogasp) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]: done

File: src/software/SfM/main_ConvertSfM_DataFormat.cpp [hunk 7]
```diff
@@ -53,24 +53,21 @@ void updateStructureWithNewUID(Landmarks &landmarks, const std::map<std::size_t,
 }
 
 /**
- * @brief Recompute the UID from the metadata of the original input images and 
- * modify the ID if it's not the same.
+ * @brief Update all viewID of a list of view by replacing them with the UID.
```
Comment (fabiencastan) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]: "by replacing them with the UID" -> not true, this function only remap the ID you give in input with oldIdToNew.

Comment (fabiencastan) [inline src/software/SfM/main_ConvertSfM_DataFormat.cpp]: Ah no, oldIdToNew is an output.
So maybe the name of the function could be more explicit: `regenerateViewsUID()`.

**Pull request merged**